include/sparrow/layout.hpp

```diff
@@ -0,0 +1,272 @@
+        const_iterator cend() const ;
+
+    private:
+        // We only use the first buffer and the bitmap.
+        array_data m_data;
+
+        pointer data();
+        const_pointer data() const;
+
+    };
+
+    /********************************************
+     * primitive_layout_iterator implementation *
+     *******************************************/
+
+    template <class T, bool is_const>
+    primitive_layout_iterator<T, is_const>::primitive_layout_iterator(pointer pointer)
+        : m_pointer(pointer)
+    {
+    }
+
+    template <class T, bool is_const>
+    auto primitive_layout_iterator<T, is_const>::dereference() const -> reference
+    {
+        return *m_pointer;
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::increment()
+    {
+        ++m_pointer;
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::decrement()
+    {
+        --m_pointer;
+    }
+
+    template <class T, bool is_const>
+    void primitive_layout_iterator<T, is_const>::advance(difference_type n)
+    {
+        m_pointer += n;
+    }
+
+    template <class T, bool is_const>
+    auto primitive_layout_iterator<T, is_const>::distance_to(const self_type& rhs) const -> difference_type
+    {
+        return rhs.m_pointer - m_pointer;
+    }
+
+    template <class T, bool is_const>
+    bool primitive_layout_iterator<T, is_const>::equal(const self_type& rhs) const
+    {
+        return distance_to(rhs) == 0;
+    }
+
+    template <class T, bool is_const>
+    bool primitive_layout_iterator<T, is_const>::less_than(const self_type& rhs) const
+    {
+        return distance_to(rhs) > 0;
+    }
+
+    /***********************************
+     * primitive_layout implementation *
+     * ********************************/
+
+    template <class T>
+    primitive_layout<T>::primitive_layout(array_data ad)
+        : m_data(ad)
+    {
+        // We only require the presence of the bitmap and the first buffer.
+        assert(m_data.buffers.size() > 0);
+        assert(m_data.length == m_data.buffers[0].size());
+        assert(m_data.length == m_data.bitmap.size());
+    }
+
+    template <class T>
+    auto primitive_layout<T>::size() const -> size_type
+    {
+        assert(m_data.buffers.size() > 0);
+        return m_data.buffers[0].size();
+    }
+
+    template <class T>
+    auto primitive_layout<T>::element(size_type i) -> reference
+    {
+        assert(pos < size());
+        return i[data()];
+    }
+
+    template <class T>
+    auto primitive_layout<T>::element(size_type i) const -> const_reference
+    {
+        assert(pos < size());
+        return i[data()];
+    }
+
+    template <class T>
+    auto primitive_layout<T>::begin() -> iterator
+    {
+        return iterator(data());
+    }
+
+    template <class T>
+    auto primitive_layout<T>::end() -> iterator
+    {
+        return iterator(data()) + self_type::size();
```

Comment:
Not important here but general note in case you didnt know: except for standard containers, you can use `{ ... }` to call constructors, in which case implicit conversions are disabled.